blockchain/src/chain_store.rs

```diff
@@ -1,21 +1,121 @@
+use std::sync::Arc;
+
 use nimiq_account::RevertInfo;
-use nimiq_block::Block;
+use nimiq_block::{Block, BlockType, EquivocationProof, MacroBody, MicroBody};
 use nimiq_blockchain_interface::{BlockchainError, ChainInfo, Direction};
 use nimiq_database::{
     declare_table,
     mdbx::{MdbxDatabase, MdbxReadTransaction, MdbxWriteTransaction, OptionalTransaction},
     traits::{Database, DupReadCursor, ReadCursor, ReadTransaction, WriteCursor, WriteTransaction},
 };
+use nimiq_database_value_derive::DbSerializable;
 use nimiq_hash::Blake2bHash;
 use nimiq_primitives::{policy::Policy, trie::trie_diff::TrieDiff};
+use nimiq_serde::{Deserialize, Serialize};
+use nimiq_transaction::{historic_transaction::HistoricTransactionData, reward::RewardTransaction};
+
+use crate::{history::interface::HistoryInterface, history_store_proxy::HistoryStoreProxy};
 
 declare_table!(HeadTable, "Head", () => Blake2bHash);
 declare_table!(ChainTable, "ChainData", Blake2bHash => ChainInfo);
-declare_table!(BlockTable, "Block", Blake2bHash => Block);
+declare_table!(PushedBlockTable, "PushedBlockTable", Blake2bHash => PushedBlock);
+declare_table!(StoredBlockTable, "StoredBlockTable", Blake2bHash => Block);
 declare_table!(HeightIndex, "HeightIndex", u32 => dup(Blake2bHash));
 declare_table!(RevertTable, "Receipts", u32 => RevertInfo);
 declare_table!(AccountsDiffTable, "AccountsDiff", Blake2bHash => TrieDiff);
 
+/// The non-header content of a block except that transactions are not stored to
+/// optimize blocks storage. This assumes that a block has been pushed and that there
+/// is history associated with this block.
+#[derive(Debug, Serialize, Deserialize, DbSerializable)]
+pub enum PushedBlock {
+    Macro,
+    Micro {
+        equivocation_proofs: Vec<EquivocationProof>,
+    },
+}
+
+impl From<Block> for PushedBlock {
+    fn from(block: Block) -> Self {
+        match block {
+            Block::Macro(_) => PushedBlock::Macro,
+            Block::Micro(micro_block) => PushedBlock::Micro {
+                equivocation_proofs: micro_block
+                    .body
+                    .expect("Block body must be present in full blockchain implementation")
+                    .equivocation_proofs,
+            },
+        }
+    }
+}
+
+impl PushedBlock {
+    fn ty(&self) -> BlockType {
+        match self {
+            PushedBlock::Macro => BlockType::Macro,
+            PushedBlock::Micro {
+                equivocation_proofs: _,
+            } => BlockType::Micro,
+        }
+    }
+    fn into_block(
+        self,
+        block: &Block,
+        history_store: &Arc<HistoryStoreProxy>,
+        txn: &MdbxReadTransaction,
+    ) -> Block {
+        assert_eq!(self.ty(), block.ty());
+        let mut block_to_return = block.clone();
```

Comment:
```suggestion
    fn populate_body(
        self,
        block: &mut Block,
        history_store: &Arc<HistoryStoreProxy>,
        txn: &MdbxReadTransaction,
    ) {
        assert_eq!(self.ty(), block.ty());
```
We can eliminate the `block.clone()` here by passing a mutable reference to partial block.